Document/0x04e-Testing-Authentication-and-Session-Management.md

```diff
@@ -47,7 +32,33 @@ For sensitive apps ("Level 2"), the MASVS adds the following:
 
 You can find details on how to test for the requirements above in the following sections.
 
-#### Testing Authentication
+#### Stateful vs. Stateless Authentication
+
+You'll usually find that the mobile app uses HTTP as the transport layer. The HTTP protocol itself is stateless, so there must be a way to associate a user's subsequent HTTP requests with that user—otherwise, the user's log in credentials would have to be sent with every request. Also, both the server and client need to keep track of user data (e.g., the user's privileges or role). This can be done in two different ways:
+
+- With *stateful* authentication, a unique session id is generated when the user logs in. In subsequent requests, this session ID serves as a reference to the user details stored on the server. The session ID is *opaque*; it doesn't contain any user data.
+
+- With *stateless* authentication, all user-identifying information is stored in a client-side token. The token can be passed to any server or micro service, eliminating the need to maintain session state on the server. Stateless authentication is often factored out to an authorization server, which produces, signs, and optionally encrypts the token upon user login.
+
+Web applications commonly use stateful authentication with a random session ID that is stored in a client-side cookie. Although mobile apps sometimes use stateful sessions in a similar fashion, stateless token-based approaches are becoming popular for a variety of reasons:
+
+- They improve scalability and performance by eliminating the need to store session state on the server.
+- Tokens enable developers to decouple authentication from the app. Tokens can be generated by an authentication server, and the authentication scheme can be changed seamlessly.
+
+As a mobile security tester, you should be familiar with both types of authentication.
+
+#### Supplementary Authentication
+
+Authentication schemes are sometimes supplemented by [passive contextual authentication](https://pdfs.semanticscholar.org/13aa/7bf53070ac8e209a84f6389bab58a1e2c888.pdf "Best Practices for Step-up Multi-factor Authentication"), which can incorporate:
+
+- Geolocation
+- IP address
+- Time of day
+- The device being used
+
+Ideally, in such a system the user's context is compared to previously recorded data to identify anomalies that might indicate account abuse or potential fraud. This process is transparent to the user, but can become a powerful deterrent to attackers.
+
```

Comment:
Given the extension, it might be a good idea to add https://webauthn.guide/ , or do we want to add an issue about it?

Reply:
Good point! I created an issue for that as this should be researched a bit more on how it can be used in iOS and Android apps (also not for the 1.2 release... ;-)
https://github.com/OWASP/owasp-mstg/issues/1345